wsmaster/che-core-api-workspace-activity/src/main/java/org/eclipse/che/api/workspace/activity/WorkspaceActivityManager.java

```diff
@@ -165,14 +191,276 @@ protected long getIdleTimeout(String wsId) {
   @ScheduleDelay(
       initialDelayParameterName = "che.workspace.activity_check_scheduler_delay_s",
       delayParameterName = "che.workspace.activity_check_scheduler_period_s")
-  private void invalidate() {
+  @VisibleForTesting
+  void validate() {
     try {
-      activityDao.findExpired(System.currentTimeMillis()).forEach(this::stopExpired);
+      stopAllExpired();
+      checkActivityRecordValidity();
     } catch (ServerException e) {
       LOG.error(e.getLocalizedMessage(), e);
     }
   }
 
+  private void stopAllExpired() throws ServerException {
+    activityDao.findExpired(clock.millis()).forEach(this::stopExpired);
+  }
+
+  private void checkActivityRecordValidity() throws ServerException {
```

Comment:
because of meaningful and useful comments this method is huge (~100 lines). Please consider using separate private methods, then it might looks like the following
```
private void checkActivityRecordValidity() throws ServerException {
  ...
  WorkspaceActivity activity = activityDao.findActivity(runningWsId);
  if (activity == null) {
    createMissingActivityRecord(...);
  } else {
    checkAndFixCreatedTime(activity);
    checkAndFixLastRunningTime(activity);
    checkAndFixExpirationTime(activity);
  }
} 

/**
* Makes sure activity record as at least 1 activity record - if it lacks last_running,
* we look at last_starting and if even that is not there we make sure there is at least
* created_time - the last_* are not mandatory, but the created_time should be present
*/
private void checkAndFixCreatedTime(...) {
   ...
}
```

Sorry, I did not fully analyze `checkActivityRecordValidity` method and I assume that it may not be so easy and nice looking as I described but please try.

Reply:
Done (to a large degree).